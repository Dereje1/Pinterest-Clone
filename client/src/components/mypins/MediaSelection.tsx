import React from 'react';
import DriveFolderUploadIcon from '@mui/icons-material/DriveFolderUpload';
import ToggleButton from '@mui/material/ToggleButton';
import ToggleButtonGroup from '@mui/material/ToggleButtonGroup';
import LinkIcon from '@mui/icons-material/Link';
import PsychologyIcon from '@mui/icons-material/Psychology';
import RestartAltIcon from '@mui/icons-material/RestartAlt';
import Typography from '@mui/material/Typography';
import UploadFileIcon from '@mui/icons-material/UploadFile';
import IconButton from '@mui/material/IconButton';
import Button from '@mui/material/Button';
import TextField from '@mui/material/TextField';

interface MediaSelectProps {
    mediaType: 'link' | 'upload' | 'AI'
    totalAiGenratedImages: number
    handleMediaChange: (e: React.SyntheticEvent, imgType: 'link' | 'upload' | 'AI') => void
  }

export function MediaSelect({
  mediaType,
  totalAiGenratedImages,
  handleMediaChange,
}: MediaSelectProps) {
  return (
    <ToggleButtonGroup
      value={mediaType}
      exclusive
      onChange={handleMediaChange}
      aria-label="text alignment"
      sx={{ marginRight: 80, color: '900' }}
    >
      <ToggleButton value="link" aria-label="link type">
        <LinkIcon />
      </ToggleButton>
      <ToggleButton value="upload" aria-label="upload type">
        <DriveFolderUploadIcon />
      </ToggleButton>
      <ToggleButton value="AI" aria-label="AI type" disabled={totalAiGenratedImages >= 5}>
        <PsychologyIcon />
      </ToggleButton>
    </ToggleButtonGroup>
  );
}

interface MediaButtonHandlerProps {
    mediaType: 'link' | 'upload' | 'AI'
    totalAiGenratedImages: number
    isError: boolean
    picPreview: string | ArrayBuffer
    description: string
    AIimageStatus: {
        generatedImage: boolean,
        generatingImage: boolean,
        _id: string | null
    }
    handleLinkImage: (e: React.SyntheticEvent) => void
    handleUploadedImage: (e: React.SyntheticEvent<HTMLDivElement>) => void
    handleAIimage: () => void
    showAIResetDialog: () => void
  }

export function MediaButtonHandler({
  mediaType,
  isError,
  picPreview,
  description,
  totalAiGenratedImages,
  AIimageStatus,
  handleLinkImage,
  handleUploadedImage,
  handleAIimage,
  showAIResetDialog,
}: MediaButtonHandlerProps) {
  const handler = {
    upload: (
      <Button
        variant="contained"
        startIcon={<UploadFileIcon />}
        sx={{ margin: '2.3vh' }}
        component="label"
        color={isError ? 'error' : 'secondary'}
      >
        {isError ? 'choose image' : 'replace image'}
        <input hidden accept="image/*" type="file" onChange={handleUploadedImage} />
      </Button>
    ),
    link: (
      <TextField
        id="pin-img-link"
        label="Paste image address here http://..."
        variant="standard"
        onChange={handleLinkImage}
        value={picPreview}
        error={isError}
        color="success"
        style={{ margin: '1.5vh', width: '100%' }}
      />
    ),
    AI: (
      <div style={{ display: 'flex', alignItems: 'center' }}>
        <Button
          variant="contained"
          startIcon={<PsychologyIcon />}
          sx={{ margin: '2.3vh' }}
          component="label"
          color="info"
          onClick={handleAIimage}
          disabled={
            AIimageStatus.generatedImage
                || AIimageStatus.generatingImage
                || !description.trim().length
                || totalAiGenratedImages >= 5
          }
        >
          {AIimageStatus.generatedImage ? 'Generated' : 'Generate Image'}
        </Button>
        <IconButton
          onClick={showAIResetDialog}
          disabled={
            !AIimageStatus.generatedImage
                || AIimageStatus.generatingImage
                || totalAiGenratedImages >= 5
          }
          disableRipple
        >
          <RestartAltIcon />
        </IconButton>
        <Typography>
          {totalAiGenratedImages}
          /5
        </Typography>
      </div>
    ),
  };

  return handler[mediaType];
}
