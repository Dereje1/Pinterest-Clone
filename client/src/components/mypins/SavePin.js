import React from 'react';
import PropTypes from 'prop-types';
import CardActions from '@mui/material/CardActions';
import SaveAltIcon from '@mui/icons-material/SaveAlt';
import IconButton from '@mui/material/IconButton';
import TextField from '@mui/material/TextField';

const SavePin = ({ isImageError, isDescriptionError, savePic }) => {
  let validation;
  if (isImageError && isDescriptionError) {
    validation = {
      text: 'Invalid image and description',
      color: '#f79f9fd9',
    };
  } else if (isImageError) {
    validation = {
      text: 'Invalid image',
      color: '#f79f9fd9',
    };
  } else if (isDescriptionError) {
    validation = {
      text: 'Invalid description',
      color: '#f79f9fd9',
    };
  } else {
    validation = {
      text: 'Save pin',
      color: '#028b02',
    };
  }
  return (
    <CardActions
      disableSpacing
    >
      <IconButton
        aria-label="Pin Image"
        onClick={savePic}
        disabled={isImageError || isDescriptionError}
      >
        <SaveAltIcon style={{ fontSize: '1.5em', color: validation.color }} />
      </IconButton>
      {
        (isImageError || isDescriptionError) && (
          <TextField
            id="pin-create-validation"
            value={validation.text}
            variant="standard"
            sx={{
              marginTop: 2,
              width: '100%',
              input: {
                fontSize: '0.75em',
                color: validation.color,
                fontWeight: 'bold',
                caretColor: 'white',
              },
            }}
            InputProps={{
              disableUnderline: true,
            }}
          />
        )
      }

    </CardActions>
  );
};

export default SavePin;

SavePin.propTypes = {
  isImageError: PropTypes.bool.isRequired,
  isDescriptionError: PropTypes.bool.isRequired,
  savePic: PropTypes.func.isRequired,
};
