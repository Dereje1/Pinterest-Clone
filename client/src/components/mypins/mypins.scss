
#mypinframe{
    margin-top: 5px;
    margin-left: 20px;
    margin-right: 20px;
    padding: 0px;
    display: flex;
    flex-direction: column;
    align-items: center;
    #username{
      font-weight: 700;
      font-family: "Helvetica";
      font-size: 36px;
      letter-spacing: -1.25px;
      line-height: 0px;
      word-spacing: -2.25px;
      color: #555;
    }
    #creatpinwrapper{
      margin: 0 auto;
      display: flex;
      flex-direction: column;
      background: rgba(0, 0, 0, 0.05);
      border-radius: 8px;
      padding: 10px;
      visibility: hidden;
      cursor: pointer;
      @media (hover: hover) {
        &:hover{
          visibility: visible;
        }
      }
      #createpin{
        height:200px;
        width:301px;
        display: flex;
        justify-content: center;
        align-items: center;
        background: rgba(0, 0, 0, 0.05);
        border-radius: 8px;
        visibility: visible;
        .fa-plus-circle{
          color:red;
          font-size:50px;
        }
      }
      #createpintext{
        color:#b5b5b5;
        font-size: 18px;
        visibility: visible;
        font-weight: 700;
        font-family: Helvetica;
      }
    }
  }