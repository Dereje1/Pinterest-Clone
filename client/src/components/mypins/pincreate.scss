.pincreate{
    position: absolute;
    top: 0;
    left: 0;
    right: 0;
    margin-left: auto;
    margin-right: auto;
    background: white;
    box-shadow: 0 5px 15px rgba(0,0,0,.5);
    border: 1px solid rgba(0,0,0,.2);
    border-radius: 6px;
    z-index: 1010;
    &.cshow{
      display: block;
      animation-name: modalshow;
      animation-duration: 1s;
      animation-timing-function: cubic-bezier(0.075, 0.82, 0.165, 1);
      animation-iteration-count: 1;
    }
    &.chide{
      display: block;
      visibility: hidden;
      animation-name: modalhide;
      animation-duration: 1s;
      animation-timing-function: cubic-bezier(0.075, 0.82, 0.165, 1);
      animation-iteration-count: .5;
    }
  }
  
  @keyframes modalshow {
    0%{
      opacity: 0;
      transform: scale(0)
    }
    100%{
      opacity: 1;
      transform: scale(1)
    }
  }
  
  @keyframes modalhide {
    0%{
      transform: scale(1);
      visibility: visible;
    }
    100%{
      transform: scale(0);
    }
  }