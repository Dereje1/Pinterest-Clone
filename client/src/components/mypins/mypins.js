// user page only for authenticated users
import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import TwitterIcon from '@mui/icons-material/Twitter';
import GoogleIcon from '@mui/icons-material/Google';
import GitHubIcon from '@mui/icons-material/GitHub';
import Typography from '@mui/material/Typography';
import Button from '@mui/material/Button';
import Dialog from '@mui/material/Dialog';
import DialogActions from '@mui/material/DialogActions';
import DialogTitle from '@mui/material/DialogTitle';
import Box from '@mui/material/Box';
import PinCreate from './pincreatemodal';
import ImageBuild from '../imagebuild/Imagebuild';
import Loading from '../imagebuild/Loading';
import RESTcall from '../../crud'; // pin CRUD
import './mypins.scss';

const getUserName = ({ service, displayName, username }) => {
  const serviceObj = {
    twitter:
  <>
    <TwitterIcon style={{ fontSize: 30, color: '#1DA1F2' }} />
    <Typography style={{ marginLeft: 15 }}>
      {`@${username}`}
    </Typography>
  </>,
    google:
  <>
    <GoogleIcon style={{ fontSize: 30, color: '#4285F4' }} />
    <Typography style={{ marginLeft: 15 }}>
      {username}
    </Typography>
  </>,
    github:
  <>
    <GitHubIcon style={{ fontSize: 30, color: '#1d7b20' }} />
    <Typography style={{ marginLeft: 15 }}>
      {`@${username}`}
    </Typography>
  </>,
  };
  return (
    <div style={{
      marginTop: 80,
      display: 'flex',
      flexDirection: 'column',
      justifyContent: 'center',
      alignItems: 'center',
    }}
    >
      <Typography variant="h4">
        {displayName}
      </Typography>
      <div style={{ display: 'flex', alignItems: 'center' }}>
        {serviceObj[service]}
      </div>
    </div>
  );
};

export class Mypins extends Component {

  constructor(props) {
    super(props);
    this.state = {
      displayPinCreate: false, // controls pin creation modal
      pinList: [], // collects the pins user owns and saved
      allPinLinks: [], // URL of all pins in DB
      showDeleteImageModal: false,
      deletableImgInfo: null,
      ready: false,
    };
  }

  async componentDidMount() {
    // get all pins and filter by owned and saved and then concatenate and set state
    const { profilePins, allPinLinks } = await RESTcall({
      address: '/api/?type=profile',
      method: 'get',
    });
    this.setState({
      pinList: profilePins,
      allPinLinks,
      ready: true,
    });
  }

  async addPic(pinJSON) { // adds a pin to the db
    // copy then add pin to db and then update client state (in that order)
    const { pinList, allPinLinks } = this.state;
    this.setState({ ready: false });
    const newPin = await RESTcall({
      address: '/api/newpin',
      method: 'post',
      payload: pinJSON,
    });
    const { owner, imgLink, originalImgLink } = newPin;
    const addedPin = {
      ...newPin,
      owner: owner.name,
      owns: true,
      hasSaved: false,
    };
    this.setState({
      pinList: [...pinList, addedPin],
      allPinLinks: [...allPinLinks, { imgLink, originalImgLink }],
      ready: true,
      displayPinCreate: false,
    });
  }

  pinForm() { // display pin creation modal
    this.setState({
      displayPinCreate: true,
    });
  }

  deletePic({ _id, imgLink, owns }) {
    const { pinList, displayPinCreate, allPinLinks } = this.state;
    if (displayPinCreate) return;
    this.setState({
      pinList: pinList.filter(p => p._id !== _id),
      showDeleteImageModal: false,
      deletableImgInfo: null,
      allPinLinks: owns ? allPinLinks.filter(links => links.imgLink !== imgLink) : allPinLinks,
    }, async () => {
      await RESTcall({
        address: owns ? `/api/${_id}` : `/api/unpin/${_id}`,
        method: owns ? 'delete' : 'put',
      });
    });
  }

  render() {
    const { user, user: { authenticated, displayName } } = this.props;
    const {
      displayPinCreate, showDeleteImageModal,
      deletableImgInfo, allPinLinks, pinList,
      ready,
    } = this.state;
    if (!authenticated) return null;
    if (!ready) return <div style={{ marginTop: 100 }}><Loading /></div>;

    return (
      <React.Fragment>
        <div>
          <div id="mypinframe">
            {getUserName(user)}
            <div
              id="creatpinwrapper"
              onClick={() => this.pinForm()}
              role="button"
              onKeyDown={() => { }}
              tabIndex={0}
            >
              <div id="createpin">
                <i className="fa fa-plus-circle" aria-hidden="true" />
              </div>
              <h3 id="createpintext">Create Pin</h3>
            </div>
          </div>
          {
            pinList.length
              ? (
                <ImageBuild
                  pinImage={null}
                  deletePin={(e) => {
                    if (e.owns) {
                      this.setState({
                        showDeleteImageModal: true,
                        deletableImgInfo: e,
                      });
                    } else {
                      this.deletePic(e);
                    }
                  }}
                  pinList={pinList}
                  ready={ready}
                  displayBrokenImage
                  displayName={displayName}
                />
              )
              : (
                <Box
                  sx={{
                    marginLeft: 'auto',
                    marginRight: 'auto',
                    marginTop: 10,
                    padding: 2,
                    backgroundColor: '#eeeeee',
                  }}
                >
                  <Typography variant="h4" style={{ textAlign: 'center' }}>
                    Welcome!
                  </Typography>
                  <Typography variant="h6" textAlign="center">
                    To get started, you can create your own pin,
                    or save an existing pin from the
                    {' '}
                    <Link to="/">home</Link>
                    {' '}
                    page
                  </Typography>
                </Box>
              )
          }

          {displayPinCreate && (
            <PinCreate
              reset={() => this.setState({ displayPinCreate: false })}
              savePin={pinJSON => this.addPic(pinJSON)}
              allPinLinks={allPinLinks}
            />
          )}
          { showDeleteImageModal && (
            <Dialog
              open={showDeleteImageModal}
              aria-labelledby="alert-dialog-title"
              aria-describedby="alert-dialog-description"
            >
              <DialogTitle id="alert-dialog-title">
                {`Permanently delete ${showDeleteImageModal && deletableImgInfo.imgDescription}?`}
              </DialogTitle>
              <DialogActions>
                <Button
                  id="cancel-delete-alert"
                  onClick={() => this.setState({
                    showDeleteImageModal: false,
                    deletableImgInfo: null,
                  })}
                >
                  Cancel
                </Button>
                <Button id="resume-delete-alert" onClick={() => this.deletePic(deletableImgInfo)} autoFocus>
                  Delete
                </Button>
              </DialogActions>
            </Dialog>
          )}
        </div>
      </React.Fragment>
    );
  }

}

const mapStateToProps = state => state;

export default connect(mapStateToProps)(Mypins);

Mypins.defaultProps = {
  user: {},
};

Mypins.propTypes = {
  // authentication info from redux
  user: PropTypes.shape(PropTypes.shape),
};
