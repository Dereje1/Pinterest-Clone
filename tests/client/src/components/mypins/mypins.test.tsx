/**
 * @jest-environment jsdom
 */
import React from 'react';
import { EnzymePropSelector, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { Mypins, mapStateToProps } from '../../../../../client/src/components/mypins/mypins';
import RESTcall from '../../../../../client/src/crud';
import { pinsStub, reduxStub } from '../../../stub';

jest.mock('../../../../../client/src/crud');
const mockedRESTcall = jest.mocked(RESTcall);

describe('The Mypins component', () => {
  let props: React.ComponentProps<typeof Mypins>;
  beforeEach(() => {
    props = {
      user: { ...reduxStub.user },
      updateDisplayName: jest.fn(),
    };
  });
  afterEach(() => {
    mockedRESTcall.mockClear();
  });

  test('landing page will render nothing if not authenticated', async () => {
    props.user.authenticated = false;
    const wrapper = shallow(<Mypins {...props} />);
    await Promise.resolve();
    expect(wrapper.isEmptyRender()).toBe(true);
  });

  test('landing page will render bubbles until ready', async () => {
    const wrapper = shallow(<Mypins {...props} />);
    await Promise.resolve();
    wrapper.setState({ ready: false });
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('landing page will render pins if found', async () => {
    const wrapper = shallow(<Mypins {...props} />);
    await Promise.resolve();
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('landing page will render message if no pins created found', async () => {
    const wrapper = shallow(<Mypins {...props} />);
    await Promise.resolve();
    wrapper.setState({ pinList: [] });
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('landing page will render message if no pins saved found', async () => {
    const wrapper = shallow(<Mypins {...props} />);
    await Promise.resolve();
    wrapper.setState({ pinList: [], displaySetting: 'saved' });
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('will cancel signal to delete/unpin if pincreate modal is showing', async () => {
    const wrapper = shallow<Mypins>(<Mypins {...props} />);
    wrapper.setState({ displayPinCreate: true });
    await Promise.resolve();
    const imageBuild: EnzymePropSelector = wrapper.find('ImageBuild');
    imageBuild.props().deletePin({ _id: 3 });
    expect(wrapper.state().pinList).toStrictEqual([pinsStub[1], pinsStub[2]]);
    expect(mockedRESTcall).toHaveBeenCalledTimes(1);
    expect(mockedRESTcall.mock.calls).toEqual([
      [{ address: '/api/mypins' }],
    ]);
  });

  test('Will display the form to create a new pin', async () => {
    const wrapper = shallow<Mypins>(<Mypins {...props} />);
    await Promise.resolve();
    expect(wrapper.state().displayPinCreate).toBe(false);
    const form = wrapper.find({ id: 'createpin' });
    form.props().onClick();
    expect(wrapper.state().displayPinCreate).toBe(true);
  });

  test('Will cancel pin delete on alert', async () => {
    const wrapper = shallow<Mypins>(<Mypins {...props} />);
    wrapper.setState({
      showDeleteImageModal: true,
      deletableImgInfo: {
        _id: '3',
        owns: true,
        imgDescription: 'deletable',
        imgLink: '',
        owner: { name: '', service: '', userId: '' },
        savedBy: [],
        hasSaved: false,
        createdAt: '',
        comments: [],
        tags: [],
      },
    });
    await Promise.resolve();
    const cancelAlert = wrapper.find({ id: 'cancel-delete-alert' });
    cancelAlert.props().onClick();
    expect(wrapper.state().showDeleteImageModal).toBe(false);
    expect(wrapper.state().deletableImgInfo).toBe(null);
  });

  test('Will delete pin on alert', async () => {
    const wrapper = shallow<Mypins>(<Mypins {...props} />);
    wrapper.setState({
      showDeleteImageModal: true,
      deletableImgInfo: {
        _id: '3',
        owns: true,
        imgDescription: 'deletable',
        imgLink: '',
        owner: { name: '', service: '', userId: '' },
        savedBy: [],
        hasSaved: false,
        createdAt: '',
        comments: [],
        tags: [],
      },
    });
    await Promise.resolve();
    const deleteAlert = wrapper.find({ id: 'resume-delete-alert' });
    deleteAlert.props().onClick();
    expect(wrapper.state().showDeleteImageModal).toBe(false);
    expect(wrapper.state().deletableImgInfo).toBe(null);
    expect(wrapper.state().pinList).toStrictEqual([pinsStub[1]]);
    expect(mockedRESTcall).toHaveBeenCalledTimes(2);
    expect(mockedRESTcall.mock.calls).toEqual([
      [{ address: '/api/mypins' }],
      [
        {
          address: '/api/3',
          method: 'delete',
        },
      ],
    ]);
  });

  test('will map redux state to the component props', () => {
    const compProps = mapStateToProps(reduxStub);
    expect(compProps).toEqual({ user: reduxStub.user });
  });
});

describe('The ImageBuild subComponent', () => {
  let props: React.ComponentProps<typeof Mypins>;
  beforeEach(() => {
    props = {
      user: { ...reduxStub.user },
      updateDisplayName: jest.fn(),
    };
  });
  afterEach(() => {
    mockedRESTcall.mockClear();
  });

  test('will recieve the pins owned as props', async () => {
    const wrapper = shallow(<Mypins {...props} />);
    await Promise.resolve();
    const imageBuild: EnzymePropSelector = wrapper.find('ImageBuild');
    const displayedPinList = imageBuild.props().pinList;
    expect(displayedPinList).toStrictEqual([pinsStub[2]]);
    expect(mockedRESTcall).toHaveBeenCalledTimes(1);
    expect(mockedRESTcall).toHaveBeenCalledWith({ address: '/api/mypins' });
  });

  test('will recieve the pins saved as props', async () => {
    const wrapper = shallow(<Mypins {...props} />);
    await Promise.resolve();
    const selector: EnzymePropSelector = wrapper.find('UserPinsSelector');
    selector.props().setDisplaySetting('saved');
    const imageBuild: EnzymePropSelector = wrapper.find('ImageBuild');
    const displayedPinList = imageBuild.props().pinList;
    expect(displayedPinList).toStrictEqual([pinsStub[1]]);
    expect(mockedRESTcall).toHaveBeenCalledTimes(1);
    expect(mockedRESTcall).toHaveBeenCalledWith({ address: '/api/mypins' });
  });

  test('will signal to unpin a pin from the db', async () => {
    const wrapper = shallow<Mypins>(<Mypins {...props} />);
    await Promise.resolve();
    const imageBuild: EnzymePropSelector = wrapper.find('ImageBuild');
    imageBuild.props().deletePin({ _id: '2' });
    expect(wrapper.state().pinList).toStrictEqual([pinsStub[2]]);
    expect(mockedRESTcall).toHaveBeenCalledTimes(2);
    expect(mockedRESTcall.mock.calls).toEqual([
      [{ address: '/api/mypins' }],
      [
        {
          address: '/api/unpin/2',
          method: 'put',
        },
      ],
    ]);
  });

  test('will signal (with modal) to delete a pin from the db', async () => {
    const wrapper = shallow<Mypins>(<Mypins {...props} />);
    await Promise.resolve();
    const instance = wrapper.instance() as Mypins;
    instance.deletePic({ _id: '3', owns: true });
    expect(wrapper.state().pinList).toStrictEqual([pinsStub[1]]);
    expect(mockedRESTcall).toHaveBeenCalledTimes(2);
    expect(mockedRESTcall.mock.calls).toEqual([
      [{ address: '/api/mypins' }],
      [
        {
          address: '/api/3',
          method: 'delete',
        },
      ],
    ]);
  });

  test('will show alert to permanently delete a pin if owner', async () => {
    const wrapper = shallow<Mypins>(<Mypins {...props} />);
    await Promise.resolve();
    const imageBuild: EnzymePropSelector = wrapper.find('ImageBuild');
    imageBuild.props().deletePin({ _id: 3, owns: true, imgDescription: 'deletable' });
    expect(wrapper.state().showDeleteImageModal).toBe(true);
    expect(wrapper.state().deletableImgInfo).toStrictEqual({ _id: 3, imgDescription: 'deletable', owns: true });
  });
});

describe('The PinCreate sub-component', () => {
  let props: React.ComponentProps<typeof Mypins>;
  beforeEach(() => {
    props = {
      user: { ...reduxStub.user },
      updateDisplayName: jest.fn(),
    };
  });
  afterEach(() => {
    mockedRESTcall.mockClear();
  });

  test('will signal to add a new pin to the db and state', async () => {
    const wrapper = shallow<Mypins>(<Mypins {...props} />);
    await Promise.resolve();
    wrapper.setState({ displayPinCreate: true });
    const createPin: EnzymePropSelector = wrapper.find('PinCreate');
    createPin.props().savePin({});
    await Promise.resolve();
    expect(wrapper.state().pinList.length).toBe(3);
    expect(wrapper.state().pinList[2]).toStrictEqual({
      savedBy: [],
      owner: { name: 'stub displayname', service: 'twitter', userId: 'a stub user id' },
      imgLink: 's3 new link',
      originalImgLink: 'original new link',
      imgDescription: 'new description',
      _id: 10,
      owns: true,
      hasSaved: false,
      createdAt: '2022-04-09T17:00:33.212Z',
      tags: null,
    });
  });

  test('will increment the total AI genrated images by the user', async () => {
    const wrapper = shallow<Mypins>(<Mypins {...props} />);
    await Promise.resolve();
    wrapper.setState({ displayPinCreate: true });
    const createPin: EnzymePropSelector = wrapper.find('PinCreate');
    expect(wrapper.state().totalAiGenratedImages).toBe(2);
    createPin.props().updateGeneratedImages();
    expect(wrapper.state().totalAiGenratedImages).toBe(3);
  });

  test('will signal to reset the display of the pin creation modal', async () => {
    const wrapper = shallow<Mypins>(<Mypins {...props} />);
    wrapper.setState({ displayPinCreate: true });
    await Promise.resolve();
    expect(wrapper.state().displayPinCreate).toBe(true);
    const createPin: EnzymePropSelector = wrapper.find('PinCreate');
    createPin.props().reset();
    expect(wrapper.state().displayPinCreate).toBe(false);
  });
});

describe('The UserInfo sub-component', () => {
  let props: React.ComponentProps<typeof Mypins>;
  beforeEach(() => {
    props = {
      user: { ...reduxStub.user },
      updateDisplayName: jest.fn(),
    };
  });
  afterEach(() => {
    mockedRESTcall.mockClear();
  });

  test('will toggle the name change form on', async () => {
    const wrapper = shallow<Mypins>(<Mypins {...props} />);
    await Promise.resolve();
    expect(wrapper.state().nameChangeFormIsShowing).toBe(false);
    const getUserInfo: EnzymePropSelector = wrapper.find('UserInfo');
    getUserInfo.props().showNameChangeForm();
    expect(wrapper.state().nameChangeFormIsShowing).toBe(true);
  });

  test('will submit a name change', async () => {
    const updatedProps = {
      ...props,
      user: {
        ...props.user,
        displayName: 'savedBy - id-2',
      },
    };
    const wrapper = shallow<Mypins>(<Mypins {...updatedProps} />);
    await Promise.resolve();
    const getUserInfo: EnzymePropSelector = wrapper.find('UserInfo');
    getUserInfo.props().submitNameChange('New name');
    await Promise.resolve();
    const [saved, owned] = wrapper.state().pinList;
    expect(mockedRESTcall).toHaveBeenCalledTimes(2);
    expect(mockedRESTcall).toHaveBeenNthCalledWith(2, {
      address: '/api/updateDisplayName',
      method: 'put',
      payload: { newDisplayName: 'New name' },
    });
    expect(owned.owner.name).toBe('New name');
    expect(saved.savedBy[0].name).toBe('New name');
    expect(props.updateDisplayName).toHaveBeenCalledWith('New name');
  });

  test('will not submit a name change if the new name is too long', async () => {
    const wrapper = shallow<Mypins>(<Mypins {...props} />);
    wrapper.setState({ nameChangeFormIsShowing: true });
    await Promise.resolve();
    const getUserInfo: EnzymePropSelector = wrapper.find('UserInfo');
    getUserInfo.props().submitNameChange('New name Is very long');
    await Promise.resolve();
    expect(mockedRESTcall).toHaveBeenCalledTimes(1);
    expect(wrapper.state().nameChangeFormIsShowing).toBe(false);
  });

  test('will not submit a name change if the new name is empty', async () => {
    const wrapper = shallow<Mypins>(<Mypins {...props} />);
    wrapper.setState({ nameChangeFormIsShowing: true });
    await Promise.resolve();
    const getUserInfo: EnzymePropSelector = wrapper.find('UserInfo');
    getUserInfo.props().submitNameChange('');
    await Promise.resolve();
    expect(mockedRESTcall).toHaveBeenCalledTimes(1);
    expect(wrapper.state().nameChangeFormIsShowing).toBe(false);
  });

  test('will not submit a name change if the new name has not changed', async () => {
    const wrapper = shallow<Mypins>(<Mypins {...props} />);
    wrapper.setState({ nameChangeFormIsShowing: true });
    await Promise.resolve();
    const getUserInfo: EnzymePropSelector = wrapper.find('UserInfo');
    getUserInfo.props().submitNameChange('tester displayName');
    await Promise.resolve();
    expect(mockedRESTcall).toHaveBeenCalledTimes(1);
    expect(wrapper.state().nameChangeFormIsShowing).toBe(false);
  });
});
