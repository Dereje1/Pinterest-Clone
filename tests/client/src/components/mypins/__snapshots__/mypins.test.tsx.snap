// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Mypins Component Mypins landing page will render bubbles until ready 1`] = `
<Loading
  marginTop={100}
/>
`;

exports[`The Mypins Component Mypins landing page will render message if no pins created found 1`] = `
<Fragment>
  <div
    style={
      {
        "alignItems": "center",
        "display": "flex",
        "flexDirection": "column",
        "justifyContent": "space-between",
        "marginLeft": "auto",
        "marginRight": "auto",
        "marginTop": 80,
        "width": 300,
      }
    }
  >
    <GetUserInfo
      nameChangeFormIsShowing={false}
      showNameChangeForm={[Function]}
      submitNameChange={[Function]}
      user={
        {
          "authenticated": true,
          "displayName": "tester displayName",
          "service": "twitter",
          "userId": "tester user Id",
          "username": "tester username",
        }
      }
    />
    <div
      style={
        {
          "marginRight": 18,
        }
      }
    >
      <UserPinsSelector
        displaySetting="created"
        setDisplaySetting={[Function]}
      />
    </div>
    <ForwardRef(Fab)
      aria-label="add"
      color="default"
      id="createpin"
      onClick={[Function]}
      sx={
        {
          "marginTop": 1,
          "zIndex": 1,
        }
      }
    >
      <Memo(ForwardRef(AddIcon))
        fontSize="large"
      />
    </ForwardRef(Fab)>
  </div>
  <ForwardRef(Box)
    sx={
      {
        "backgroundColor": "#eeeeee",
        "marginLeft": "auto",
        "marginRight": "auto",
        "marginTop": 10,
        "padding": 2,
      }
    }
  >
    <ForwardRef(Typography)
      style={
        {
          "textAlign": "center",
        }
      }
      variant="h4"
    >
      Welcome!
    </ForwardRef(Typography)>
    <ForwardRef(Typography)
      textAlign="center"
      variant="h6"
    >
      You have not created any pins, you can create some by clicking on the button above
    </ForwardRef(Typography)>
  </ForwardRef(Box)>
</Fragment>
`;

exports[`The Mypins Component Mypins landing page will render message if no pins saved found 1`] = `
<Fragment>
  <div
    style={
      {
        "alignItems": "center",
        "display": "flex",
        "flexDirection": "column",
        "justifyContent": "space-between",
        "marginLeft": "auto",
        "marginRight": "auto",
        "marginTop": 80,
        "width": 300,
      }
    }
  >
    <GetUserInfo
      nameChangeFormIsShowing={false}
      showNameChangeForm={[Function]}
      submitNameChange={[Function]}
      user={
        {
          "authenticated": true,
          "displayName": "tester displayName",
          "service": "twitter",
          "userId": "tester user Id",
          "username": "tester username",
        }
      }
    />
    <div
      style={
        {
          "marginRight": 18,
        }
      }
    >
      <UserPinsSelector
        displaySetting="saved"
        setDisplaySetting={[Function]}
      />
    </div>
    <ForwardRef(Fab)
      aria-label="add"
      color="default"
      id="createpin"
      onClick={[Function]}
      sx={
        {
          "marginTop": 1,
          "zIndex": 1,
        }
      }
    >
      <Memo(ForwardRef(AddIcon))
        fontSize="large"
      />
    </ForwardRef(Fab)>
  </div>
  <ForwardRef(Box)
    sx={
      {
        "backgroundColor": "#eeeeee",
        "marginLeft": "auto",
        "marginRight": "auto",
        "marginTop": 10,
        "padding": 2,
      }
    }
  >
    <ForwardRef(Typography)
      textAlign="center"
      variant="h6"
    >
      You have no pins saved, you can save some from the
       
      <Link
        to="/"
      >
        home
      </Link>
       
      page
    </ForwardRef(Typography)>
  </ForwardRef(Box)>
</Fragment>
`;

exports[`The Mypins Component Mypins landing page will render pins if found 1`] = `
<Fragment>
  <div
    style={
      {
        "alignItems": "center",
        "display": "flex",
        "flexDirection": "column",
        "justifyContent": "space-between",
        "marginLeft": "auto",
        "marginRight": "auto",
        "marginTop": 80,
        "width": 300,
      }
    }
  >
    <GetUserInfo
      nameChangeFormIsShowing={false}
      showNameChangeForm={[Function]}
      submitNameChange={[Function]}
      user={
        {
          "authenticated": true,
          "displayName": "tester displayName",
          "service": "twitter",
          "userId": "tester user Id",
          "username": "tester username",
        }
      }
    />
    <div
      style={
        {
          "marginRight": 18,
        }
      }
    >
      <UserPinsSelector
        displaySetting="created"
        setDisplaySetting={[Function]}
      />
    </div>
    <ForwardRef(Fab)
      aria-label="add"
      color="default"
      id="createpin"
      onClick={[Function]}
      sx={
        {
          "marginTop": 1,
          "zIndex": 1,
        }
      }
    >
      <Memo(ForwardRef(AddIcon))
        fontSize="large"
      />
    </ForwardRef(Fab)>
  </div>
  <ImageBuild
    deletePin={[Function]}
    displayBrokenImage={true}
    pinImage={false}
    pinList={
      [
        {
          "_id": "3",
          "comments": [],
          "createdAt": "2020-12-25T00:00:00.000+00:00",
          "hasSaved": false,
          "imgDescription": "imgDescription id-3",
          "imgLink": "imgLink id-3",
          "owner": {
            "name": "owner id-3",
            "service": "github",
            "userId": "3",
          },
          "owns": true,
          "savedBy": [
            {
              "name": "savedBy - id-3",
              "service": "google",
              "userId": "3",
            },
            {
              "name": "savedBy - id-3",
              "service": "github",
              "userId": "3",
            },
          ],
          "tags": [],
        },
      ]
    }
    ready={true}
    user={
      {
        "authenticated": true,
        "displayName": "tester displayName",
        "service": "twitter",
        "userId": "tester user Id",
        "username": "tester username",
      }
    }
  />
</Fragment>
`;
